test/e2e/reconciler/apiexportendpointslice/apiexportendpointslice_test.go

```diff
@@ -0,0 +1,329 @@
+	require.True(t, apierrors.IsForbidden(err), "no error creating APIExportEndpointSlice (admission should have declined it)")
+
+	t.Logf("Creating the missing APIExport")
+	exportClient := kcpClusterClient.ApisV1alpha1().APIExports()
+	_, err = exportClient.Cluster(exportClusterPath).Create(ctx, export, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating APIExport")
+
+	t.Logf("Retrying to create the APIExportEndpointSlice after the APIExport has been created")
+	require.Eventually(t, func() bool {
+		slice, err = sliceClient.Cluster(partitionClusterPath).Create(ctx, slice, metav1.CreateOptions{})
+		return err == nil
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected APIExportEndpointSlice creation to succeed")
+
+	require.Eventually(t, func() bool {
+		slice, err = kcpClusterClient.Cluster(partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return conditions.IsTrue(slice, apisv1alpha1.APIExportValid)
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected valid APIExport")
+
+	t.Logf("Adding a Partition to the APIExportEndpointSlice")
+	slice.Spec.Partition = partition.Name
+	_, err = kcpClusterClient.Cluster(partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Update(ctx, slice, metav1.UpdateOptions{})
+	require.NoError(t, err, "error updating APIExportEndpointSlice")
+
+	require.Eventually(t, func() bool {
+		slice, err = kcpClusterClient.Cluster(partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return conditions.IsFalse(slice, apisv1alpha1.PartitionValid) && conditions.GetReason(slice, apisv1alpha1.PartitionValid) == apisv1alpha1.PartitionInvalidReferenceReason
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected missing Partition")
+	require.True(t, len(slice.Status.APIExportEndpoints) == 0, "not expecting any endpoint")
+
+	t.Logf("Creating the missing Partition")
+	partitionClient := kcpClusterClient.TopologyV1alpha1().Partitions()
+	_, err = partitionClient.Cluster(partitionClusterPath).Create(ctx, partition, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating Partition")
+
+	require.Eventually(t, func() bool {
+		s, err := kcpClusterClient.Cluster(partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return conditions.IsTrue(s, apisv1alpha1.PartitionValid)
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expected valid Partition")
+
+	t.Logf("Checking that no endpoints have been populated")
+	require.Eventually(t, func() bool {
+		s, err := kcpClusterClient.Cluster(partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return len(s.Status.APIExportEndpoints) == 0
```

Comment:
what is the point of doing this call in `require.Eventually` since it will stop after the first iteration?

Reply:
changed